src/off-chain/historical-data/cowEndpointSurplus.py

```diff
@@ -0,0 +1,258 @@
+
+        self.solver_dict = {
+            "1Inch": [0, 0],
+            "Raven": [0, 0],
+            "0x": [0, 0],
+            "PLM": [0, 0],
+            "Quasilabs": [0, 0],
+            "Otex": [0, 0],
+            "SeaSolver": [0, 0],
+            "Laertes": [0, 0],
+            "ParaSwap": [0, 0],
+            "BalancerSOR": [0, 0],
+            "BaselineSolver": [0, 0],
+            "Legacy": [0, 0],
+            "NaiveSolver": [0, 0],
+            "DMA": [0, 0],
+            "CowDexAg": [0, 0],
+            "Barter": [0, 0],
+        }
+
+    def total_surplus(
+        self, start_block=None, end_block=None, settlement_hash=None, file_name=None
+    ) -> None:
+        self.file_name = file_name
+
+        if settlement_hash is not None:
+            self.get_order_surplus(settlement_hash)
+
+        elif start_block is not None and end_block is not None:
+            # Etherscan endpoint call for settlements between start and end block
+            try:
+                etherscan_url = f"https://api.etherscan.io/api?module=account&action=txlist&address=0x9008D19f58AAbD9eD0D60971565AA8510560ab41&startblock={start_block}&endblock={end_block}&sort=desc&apikey={directory.ETHERSCAN_KEY}"
+                # all "result" go into results (based on API return value names from docs)
+                settlements = json.loads((requests.get(etherscan_url)).text)["result"]
+
+                # loop over all settlements and get surplus difference for each
+                for settlement in settlements:
+                    self.get_order_surplus(settlement["hash"])
+                self.statistics_output(start_block, end_block)
+            except ValueError:
+                print("etherscan error.")
+
+    def get_order_surplus(self, settlement_hash: str) -> None:
+        # Once we have settlement transaction hashes, call competition endpoint to get solver data
+        endpoint_url = f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlement_hash}"
+        json_competition_data = requests.get(endpoint_url)
+
+        if json_competition_data.status_code == 200:
+            competition_data = json.loads(json_competition_data.text)
+            winning_solution = competition_data["solutions"][-1]
+            winning_solver = winning_solution["solver"]
+            winning_orders = winning_solution["orders"]
+
+            for individual_win_order in winning_orders:
+                self.solver_dict[winning_solver][0] += 1
+                self.total_orders = self.total_orders + 1
+                individual_win_order_id = individual_win_order["id"]
+                order_data_url = f"https://api.cow.fi/mainnet/api/v1/orders/{individual_win_order_id}"
+                json_order = requests.get(order_data_url)
+                if json_order.status_code == 200:
+                    individual_order_data = json.loads(json_order.text)
+                    if individual_order_data["isLiquidityOrder"] is False:
```

Comment:
Personal preference, just to avoid pushing the code more and more to the right due to indentation:

```
if order is liquidity:
    continue

```

Reply:
Also note that you have already counted such orders in the` total_orders` accounting.